mysql/contrib/common.sh

```diff
@@ -0,0 +1,165 @@
+  echo "  MYSQL_LOWER_CASE_TABLE_NAMES (default: 0)"
+  echo "  MYSQL_MAX_CONNECTIONS (default: 151)"
+  echo "  MYSQL_FT_MIN_WORD_LEN (default: 4)"
+  echo "  MYSQL_FT_MAX_WORD_LEN (default: 20)"
+  echo "  MYSQL_AIO (default: 1)"
+  exit 1
+}
+
+function validate_variables() {
+  # Check basic sanity of specified variables
+  if [[ -v MYSQL_USER && -v MYSQL_PASSWORD && -v MYSQL_DATABASE ]]; then
+    [[ "$MYSQL_USER"     =~ $mysql_identifier_regex ]] || usage "Invalid MySQL username"
+    [ ${#MYSQL_USER} -le 16 ] || usage "MySQL username too long (maximum 16 characters)"
+    [[ "$MYSQL_PASSWORD" =~ $mysql_password_regex   ]] || usage "Invalid password"
+    [[ "$MYSQL_DATABASE" =~ $mysql_identifier_regex ]] || usage "Invalid database name"
+    [ ${#MYSQL_DATABASE} -le 64 ] || usage "Database name too long (maximum 64 characters)"
+    user_specified=1
+  fi
+
+  if [ -v MYSQL_ROOT_PASSWORD ]; then
+    [[ "$MYSQL_ROOT_PASSWORD" =~ $mysql_password_regex ]] || usage "Invalid root password"
+    root_specified=1
+  fi
+
+  # Either combination of user/pass/db or root password is ok
+  if [[ "${user_specified:-0}" == "0" && "${root_specified:-0}" == "0" ]]; then
+    usage
+  fi
+
+  # Specifically check of incomplete specification
+  if [[ -v MYSQL_USER || -v MYSQL_PASSWORD || -v MYSQL_DATABASE ]] && \
+     [[ "${user_specified:-0}" == "0" ]]; then
+    usage
+  fi
+}
+
+# Make sure env variables don't propagate to mysqld process.
+function unset_env_vars() {
+  unset MYSQL_USER MYSQL_PASSWORD MYSQL_DATABASE MYSQL_ROOT_PASSWORD
+}
+
+# Poll until MySQL responds to our ping.
+function wait_for_mysql() {
+  pid=$1 ; shift
+
+  while [ true ]; do
+    if [ -d "/proc/$pid" ]; then
+      mysqladmin --socket=/tmp/mysql.sock ping &>/dev/null && return 0
+    else
+      return 1
+    fi
+    echo "Waiting for MySQL to start ..."
+    sleep 1
+  done
+}
+
+function start_local_mysql() {
+  # Now start mysqld and add appropriate users.
+  echo 'Starting local mysqld server ...'
+  /usr/libexec/mysqld \
+    --defaults-file=$MYSQL_DEFAULTS_FILE \
+    --skip-networking --socket=/tmp/mysql.sock "$@" &
+  mysql_pid=$!
+  wait_for_mysql $mysql_pid
+}
+
+# Initialize the MySQL database (create user accounts and the initial database)
+function initialize_database() {
+  echo 'Running mysql_install_db ...'
+  # Using --rpm since we need mysql_install_db behaves as in RPM
+  mysql_install_db --rpm --datadir=$MYSQL_DATADIR
+  start_local_mysql "$@"
+
+  [ -v MYSQL_RUNNING_AS_SLAVE ] && return
+
+  # Do not care what option is compulsory here, just create what is specified
+  if [ -v MYSQL_USER ]; then
+mysql $mysql_flags <<EOSQL
+    CREATE USER '${MYSQL_USER}'@'%' IDENTIFIED BY '${MYSQL_PASSWORD}';
+EOSQL
+  fi
+
+  if [ -v MYSQL_DATABASE ]; then
+    mysqladmin $admin_flags create "${MYSQL_DATABASE}"
+
+    if [ -v MYSQL_USER ]; then
+mysql $mysql_flags <<EOSQL
+      GRANT ALL ON \`${MYSQL_DATABASE}\`.* TO '${MYSQL_USER}'@'%' ;
+      FLUSH PRIVILEGES ;
```

Comment:
First comment I originally intended to make, to improve this - FLUSH PRIVILEGES should not be necessary here.  According to rant: http://dbahire.com/stop-using-flush-privileges/